### PR TITLE
feat: batch isolation in relaxation

### LINKED ISSUE #94: Batched relaxation shares one LBFGS state across independent systems, degrading convergence
[shared_lbfgs_bug.py](https://github.com/user-attachments/files/27264118/shared_lbfgs_bug.py)

## Summary

When `relax_mlff` is given multiple independent structures via `inp_files`, all molecules share a single LBFGS optimizer state over the concatenated position vector. The L-BFGS memory vectors (`s`, `y`) mix curvature information from different potential energy surfaces, corrupting the inverse Hessian approximation and severely degrading per-system convergence.

We believe this is a promising framework and are submitting this issue (and an accompanying PR with a proposed fix) in the spirit of being helpful. This was discovered during benchmarking by Sam Blau and an AI assistant (Claude, via [Axon](https://www.mirrorphysics.com) by Mirror Physics) while evaluating kUPS for batched molecular geometry optimization with jaxified UMA.

## Reproduction

The root cause can be demonstrated with a pure-NumPy synthetic test (no kUPS or ML model needed) — see `examples/shared_lbfgs_bug.py` in the accompanying PR, and the file is also attached here. Two quadratic systems with different curvatures share one LBFGS vs having independent LBFGS instances:

```
After 50 steps:
  System A — Independent: 2.14e-11, Shared: 4.98e-02 (2327126162x worse)
  System B — Independent: 3.42e-02, Shared: 3.04e-02
```

## Impact with real ML potentials

We benchmarked 25 drug-like molecules with jaxified UMA-s-1p2 (OMol task):

| Method | Converged (fmax < 0.01 eV/Å) | Wall time |
|--------|-------------------------------|-----------|
| ASE LBFGS (sequential, per-molecule) | 22/25 | 287s |
| kUPS `relax_mlff` (shared LBFGS) | 1/25 | 145s |
| **Per-molecule vmapped LBFGS** (proposed fix) | **22/25** | **19s** |

The fix uses `jax.vmap` over per-molecule optimizer states while keeping the batched model forward pass. Step counts match ASE exactly (e.g., tafamidis: 19 steps in both).

## Root cause

In `relax_mlff.py`, the optimizer is initialized over the concatenated position vector of all systems:

```python
opt_state = opt_init(
    (particles.data.positions, systems.data.unitcell.lattice_vectors)
)
```

The `scale_by_ase_lbfgs` transform stores `diff_params_memory` and `diff_updates_memory` arrays over this concatenated space. The two-loop LBFGS recursion computes `s·y` dot products that mix gradient information from independent potential energy surfaces.

## Suggested fix

After the batched gradient computation, split gradients by system, pad to `(B, max_atoms, 3)`, and use `jax.vmap(optimizer.update)` over per-molecule optimizer states. The vmapped LBFGS kernel compiles in <1s and runs in <1ms for 25 molecules — the overhead is negligible.

We have a working prototype (see accompanying PR) demonstrating the approach.

### PR DESCRIPTION
**Pull request by n-gao**
We introduce a new Optimizer class that may wrap optax optimizers. It's key distinction is the addition of an index_prefix during state initialization that allows us to define the batching behavior for different data. We reimplement all existing modules in this new framework.  
  
For better composability, we separated the last step of FIRE (norm clipping) into ClipByGlobalNorm.  
  
co-authored by @jonkhler   

fixes #94
supersedes #95

### CODE REVIEW

**Comment by n-gao** [comment]
* **#111** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/109?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/109?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#107** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/107?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#106** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/106?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#105** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/105?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#104** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/104?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by jonkhler — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **May 11, 1:59 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/109).
